url70997c83aa_Xilaew_risiko_tgz-pJ8-risiko_XMI2XMLtransformerJ8/src/risiko/main/MainNio.java

```diff
@@ -0,0 +1,146 @@
+package risiko.main;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.util.Iterator;
+
+import risiko.Engine;
+
+public class MainNio implements Runnable {
+	private final int port;
+	private ServerSocketChannel ssc;
+	private Selector selector;
+	private Engine engine;
+	private ByteBuffer buf = ByteBuffer.allocate(256);
+
+	MainNio(int port) throws IOException {
+		this.port = port;
+		this.ssc = ServerSocketChannel.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when AbstractSelectableChannel.configureBlocking throws an exception, the *MainNio* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *ssc* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).